packages/okta-auth-js/lib/token.js

```diff
@@ -760,16 +760,37 @@ function parseFromUrl(sdk, url) {
     });
 }
 
-function getUserInfo(sdk, accessTokenObject) {
+async function getUserInfo(sdk, accessTokenObject, idTokenObject) {
+  // If token objects were not passed, attempt to read from the TokenManager
+  if (!accessTokenObject) {
+    accessTokenObject = await sdk.tokenManager.get('accessToken');
+  }
+  if (!idTokenObject) {
+    idTokenObject = await sdk.tokenManager.get('idToken');
+  }
+
   if (!accessTokenObject ||
       (!oauthUtil.isToken(accessTokenObject) && !accessTokenObject.accessToken && !accessTokenObject.userinfoUrl)) {
     return Promise.reject(new AuthSdkError('getUserInfo requires an access token object'));
   }
+
+  if (!idTokenObject ||
+    (!oauthUtil.isToken(idTokenObject) && !idTokenObject.idToken)) {
+    return Promise.reject(new AuthSdkError('getUserInfo requires an ID token object'));
+  }
+
   return http.httpRequest(sdk, {
     url: accessTokenObject.userinfoUrl,
     method: 'GET',
     accessToken: accessTokenObject.accessToken
   })
+  .then(userInfo => {
+    // Only return the userinfo response if subjects match to mitigate token substitution attacks
+    if (userInfo.sub === idTokenObject.claims.sub) {
+      return userInfo;
+    }
+    return Promise.reject(new AuthSdkError('getUserInfo request was rejected due to token mismatch'));
```

Comment:
Is this most likely a MITM attack that we should highlight more, or are there more innocent causes?

Reply:
Moving to PKCE default and enforcing HTTPS origin, I think MitM attacks would be very hard to pull off, they would mostly affect unencrypted connections. There are a lot of bad things a man in the middle could do, including intercept your username and password. This specific safeguard (validating userinfo against idToken) already exists in our SDKs. By moving it here we can remove this logic in our SDKs.

We are also adding in this 3.0 release validation of idToken against accessToken (at_hash). This I think would also defend against some types of attacks.

Overall, I would say it is not worth calling out except to reiterate our strong recommendation to use HTTPS connections.